sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -906,3 +907,36 @@ case class Deduplicate(
 
   override def output: Seq[Attribute] = child.output
 }
+
+/** A logical plan for setting a barrier of analysis */
+case class AnalysisBarrier(child: LogicalPlan) extends UnaryNode {
```

Comment:
A simple way is to make `AnalysisBarrier` a leaf node, so that we won't transform through the wrapped plan.